### PR TITLE
Block processing refactoring (switch to new BlockApi, Casper cleanup)

### LINKED ISSUE #3659: Refactor BlockAPI
## Overview

To include new multi-parent finalization we need to refactor/delete a lot of existing code.

As part of this change we need to simplify current `BlockAPI` implementation.

## Design

1. Create `BlockAPI` trait with existing methods currently defined on the BlockAPI object
    - for now it can be in models project (because we have generated protobuf there!) in `coop.rchain.api`
1. Create implementation which should only accept RuntimeManager and BlockDagStorage (BlockDagRepresentation)
    - for all other dependencies just add TODO comment with fake data
1. Instantiate new BlockAPI implementation in `Setup` and use where is needed like gRPC servers and WebAPI

## Additional changes

Currently, `addDeploy` method is defined on MultiParentCasperImpl which is used in BlockAPI. This should be moved directly to new BlockAPI implementation and DeloyStorage should be accessible in BlockDagStorage.

### PR DESCRIPTION
**Pull request by tgrospic**
## Overview

This PR is part of the refactoring and cleanup for the new finalizer.

The main part is removal of most methods from Casper and MultiParentCasper interfaces and preparation for complete removal.

Also old block API file is removed and switch to new BlockAPI implementation is completed without dependency to _Casper_ and config parameters are now provided to BlockAPI class and not individual methods.

Some unused block statuses and errors are removed. E.g. wrong block hash cannot be part of slashing because block identity can be shadowed, parents relation will not be used anymore, wrong shard shard ID will be ignored...

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: casper/src/main/scala/coop/rchain/casper/storage/RNodeKeyValueStoreManager.scala [hunk 1]
```diff
@@ -59,28 +50,16 @@ object RNodeKeyValueStoreManager {
...
-          (Db("rspace-roots"), rspaceHistoryEnvConfig),
-          (Db("rspace-cold"), rspaceColdEnvConfig)
-        )
-      } else
-        // Legacy config has the same database name for all maps
-        Seq(
-          (Db("rspace-history", nameOverride = "db".some), legacyEnvConfig("history")),
-          (Db("rspace-roots", nameOverride = "db".some), legacyEnvConfig("roots")),
-          (Db("rspace-cold", nameOverride = "db".some), legacyEnvConfig("cold")),
-          (Db("rspace-channels", nameOverride = "db".some), legacyEnvConfig("channels"))
```
Comment (stanislavlyalin) [inline casper/src/main/scala/coop/rchain/casper/storage/RNodeKeyValueStoreManager.scala:83]: Is this legasy data used for RNode compability with old data formats? Why it is not important now?

File: casper/src/main/scala/coop/rchain/casper/Validate.scala [hunk 2]
```diff
@@ -154,23 +154,15 @@ object Validate {
...
       shardId: String,
       expirationThreshold: Int
   ): F[ValidBlockProcessing] =
     (for {
-      _ <- EitherT.liftF(Span[F].mark("before-block-hash-validation"))
-      _ <- EitherT(Validate.blockHash(block))
       _ <- EitherT.liftF(Span[F].mark("before-timestamp-validation"))
       _ <- EitherT(Validate.timestamp(block))
-      _ <- EitherT.liftF(Span[F].mark("before-shard-identifier-validation"))
-      _ <- EitherT(Validate.shardIdentifier(block, shardId))
```
Comment (stanislavlyalin) [inline casper/src/main/scala/coop/rchain/casper/Validate.scala:173]: Why `shardId` validation for block was removed?

File: casper/src/main/scala/coop/rchain/casper/Casper.scala [hunk 3]
```diff
@@ -52,7 +52,6 @@ trait Casper[F[_]] {
   def dagContains(hash: BlockHash): F[Boolean]
   def bufferContains(hash: BlockHash): F[Boolean]
   def deploy(d: Signed[DeployData]): F[Either[DeployError, DeployId]]
-  def getApprovedBlock: F[BlockMessage]
```
Comment (stanislavlyalin) [inline casper/src/main/scala/coop/rchain/casper/Casper.scala:56]: I missed your conversation with Anton. Why approved block can be removed? Is ApprovedStore still necessary?

File: casper/src/main/scala/coop/rchain/casper/engine/Running.scala [hunk 4]
```diff
@@ -51,42 +48,33 @@ object Running {
     * To mitigate this issue we can update fork choice tips if current fork-choice tip has old timestamp,
     * which means node does not propose new blocks and no new blocks were received recently.
     */
-  def updateForkChoiceTipsIfStuck[F[_]: Sync: CommUtil: Log: Time: BlockStore: BlockDagStorage: EngineCell](
+  def updateForkChoiceTipsIfStuck[F[_]: Sync: CommUtil: Log: Time: BlockStore: BlockDagStorage](
       delayThreshold: FiniteDuration
   ): F[Unit] =
     for {
-      engine <- EngineCell[F].read
-      _ <- engine.withCasper(
```
Comment (stanislavlyalin) [inline casper/src/main/scala/coop/rchain/casper/engine/Running.scala:59]: Now `withCasper` is used only in tests. Can we fully remove it?

File: casper/src/main/scala/coop/rchain/casper/Validate.scala [hunk 5]
```diff
@@ -420,21 +408,6 @@ object Validate {
...
-    if (b.shardId == shardId) {
-      BlockStatus.valid.asRight[BlockError].pure
-    } else {
-      for {
-        _ <- Log[F].warn(
-              ignore(b, s"got shard identifier ${b.shardId} while $shardId was expected.")
-            )
-      } yield BlockStatus.invalidShardId.asLeft[ValidBlock]
-    }
-
```
Comment (tgrospic) [inline casper/src/main/scala/coop/rchain/casper/Validate.scala:437]: Wrong shard will be ignored block.

**Review by stanislavlyalin — COMMENTED** [review]
Less code - fewer potential problems :slightly_smiling_face:

File: casper/src/test/scala/coop/rchain/casper/api/PreviewPrivateNameTest.scala [hunk 6]
```diff
@@ -1,55 +0,0 @@
-package coop.rchain.casper.api
```
Comment (tgrospic) [inline casper/src/test/scala/coop/rchain/casper/api/PreviewPrivateNameTest.scala:1]: We will remove private names from API completely so this test becomes irrelevant.

File: casper/src/test/scala/coop/rchain/casper/batch2/SingleParentCasperSpec.scala [hunk 7]
```diff
@@ -42,35 +38,4 @@ class SingleParentCasperSpec extends FlatSpec with Matchers with Inspectors {
...
-          validateResult <- {
-            import n1._
-            implicit val s  = Sync[Task]
-            implicit val sp = NoopSpan[Task]
-            n1.casperEff.getSnapshot >>=
-              (snap => Validate.parents(dualParentB3, n1.genesis, snap))
-          }
-        } yield validateResult shouldBe BlockStatus.invalidParents.asLeft[ValidBlock]
-    }
-  }
```
Comment (tgrospic) [inline casper/src/test/scala/coop/rchain/casper/batch2/SingleParentCasperSpec.scala:75]: Parent blocks relation will be removed.

File: casper/src/test/scala/coop/rchain/casper/batch2/ValidateTest.scala [hunk 8]
```diff
@@ -601,85 +520,6 @@ class ValidateTest
...
-            ) shouldBeF true
-        blockId7 <- blockDagStorage.lookupByIdUnsafe(7)
-        dag      <- blockDagStorage.getRepresentation
-        _ <- Validate.justificationFollows[Task](
-              blockId7
-            ) shouldBeF Left(InvalidFollows)
-        _      = log.warns.size shouldBe 1
-        result = log.warns.forall(_.contains("do not match the bonded validators")) shouldBe true
-      } yield result
-  }
```
Comment (tgrospic) [inline casper/src/test/scala/coop/rchain/casper/batch2/ValidateTest.scala:681]: Justification validation will be included in finalization code.

File: casper/src/test/scala/coop/rchain/casper/helper/TestNode.scala [hunk 9]
```diff
@@ -341,44 +317,16 @@ case class TestNode[F[_]: Timer](
...
-    */
-  def printVisualizeDagUrl(startBlockNumber: Int): F[Unit] =
-    for {
-      dot <- visualizeDag(startBlockNumber)
-      // Java's URLEncode encodes ' ' as '+' instead of '%20', see https://stackoverflow.com/a/5330239/52142
-      urlEncoded = URLEncoder.encode(dot, "UTF-8").replaceAll("\\+", "%20")
-      _ <- Sync[F].delay {
-            println(s"DAG @ $name: https://dreampuf.github.io/GraphvizOnline/#" + urlEncoded)
-          }
-    } yield ()
```
Comment (tgrospic) [inline casper/src/test/scala/coop/rchain/casper/helper/TestNode.scala:381]: This is not used.

File: casper/src/test/scala/coop/rchain/casper/helper/TestNode.scala [hunk 10]
```diff
@@ -506,6 +454,26 @@ object TestNode {
...
+        // Required to enable protection from re-submitting duplicate deploys
+        deployLifespan = 50,
+        casperVersion = 1,
+        configVersion = 1,
+        bondMinimum = 0,
+        bondMaximum = Long.MaxValue,
+        epochLength = 10000,
+        quarantineLength = 20000,
+        minPhloPrice = 1
+      )
```
Comment (tgrospic) [inline casper/src/test/scala/coop/rchain/casper/helper/TestNode.scala:465]: For now shard config is created inside TestNode, it will probably be changed in subsequent PRs.

File: .github/workflows/continuous-integration.yml [hunk 11]
```diff
@@ -286,10 +286,10 @@ jobs:
 #          - test_finalization
           - test_genesis_ceremony
           - test_propose
-          - test_slash_invalid_block_hash
+#          - test_slash_invalid_block_hash
           - test_slash_invalid_block_number
           - test_slash_invalid_block_seq
-          - test_slash_justification_not_correct
+#          - test_slash_justification_not_correct
```
Comment (tgrospic) [inline .github/workflows/continuous-integration.yml:292]: I'm not sure what is wrong for these tests to fail. We will fix this after refactoring of block validation.

File: casper/src/main/scala/coop/rchain/casper/Validate.scala (hunk 2 above)
Comment (nzpr) [reply casper/src/main/scala/coop/rchain/casper/Validate.scala:173]: https://github.com/rchain/rchain/pull/3672/commits/fd688400030326ac3e1f8173c735455d20a3fe63#r860065262 its should not slashable offence but just a reason to drop block on the floor I guess. This should be on block receiver not here.

File: casper/src/test/scala/coop/rchain/casper/addblock/MultiParentCasperAddBlockSpec.scala [hunk 12]
```diff
@@ -483,7 +483,8 @@ class MultiParentCasperAddBlockSpec extends FlatSpec with Matchers with Inspecto
   it should "succeed at slashing" in effectTest {
     TestNode.networkEff(genesis, networkSize = 3).use { nodes =>
       for {
-        deployData   <- ConstructDeploy.basicDeployData[Effect](0)
+        deployData <- ConstructDeploy
+                       .basicDeployData[Effect](0, shardId = genesis.genesisBlock.shardId)
```
Comment (nzpr) [inline casper/src/test/scala/coop/rchain/casper/addblock/MultiParentCasperAddBlockSpec.scala]: I'm not sure - why this change is required? shard validation removed so this should not make any difference?

File: casper/src/main/scala/coop/rchain/casper/BlockStatus.scala [hunk 13]
```diff
@@ -13,12 +13,11 @@ object BlockStatus {
   def invalidBlockNumber: BlockError       = InvalidBlock.InvalidBlockNumber
   def invalidRepeatDeploy: BlockError      = InvalidBlock.InvalidRepeatDeploy
   def invalidSequenceNumber: BlockError    = InvalidBlock.InvalidSequenceNumber
-  def invalidShardId: BlockError           = InvalidBlock.InvalidShardId
+  def invalidDeployShardId: BlockError     = InvalidBlock.InvalidDeployShardId
```
Comment (nzpr) [inline casper/src/main/scala/coop/rchain/casper/BlockStatus.scala:16]: Why it's `invalidDeployShardId ` shardID still supposed to be related to a block? Or you want it per deploy?

File: casper/src/main/scala/coop/rchain/casper/blocks/BlockProcessor.scala [hunk 14]
```diff
@@ -19,7 +19,7 @@ import coop.rchain.shared.Log
   * Logic for processing incoming blocks
   * Blocks created by node itself are not held here, but in Proposer.
   */
-class BlockProcessor[F[_]: Concurrent](
+class BlockProcessor[F[_]: Concurrent: BlockDagStorage: CasperBufferStorage](
```
Comment (nzpr) [inline casper/src/main/scala/coop/rchain/casper/blocks/BlockProcessor.scala]: Will this BlockProcessor be removed completely and replaced with BlockReceiver?
I'm curious why shardOfInterest validation is removed a bit lower?

File: casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala [hunk 15]
```diff
@@ -385,4 +338,51 @@ object MultiParentCasperImpl {
...
+      blockMessage <- dag.lastFinalizedBlockUnsafe.flatMap(BlockStore[F].getUnsafe)
+    } yield blockMessage
+
+  def handleValidBlock[F[_]: Sync: BlockDagStorage: BlockStore: CasperBufferStorage](
+      block: BlockMessage
+  ): F[DagRepresentation] =
+    for {
+      updatedDag <- BlockDagStorage[F].insert(block, invalid = false)
+      _          <- CasperBufferStorage[F].remove(block.blockHash)
+      _          <- lastFinalizedBlock
```
Comment (nzpr) [inline casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala:274]: This were supposed to update finalization (thats why output is discarded), but now it looks like it is just reading of a value. It's fine for now as finalization is to be implemented, but just to want to point at this.

File: casper/src/test/scala/coop/rchain/casper/helper/TestNode.scala (hunk 9 above)
Comment (nzpr) [reply casper/src/test/scala/coop/rchain/casper/helper/TestNode.scala:381]: Not sure wdy mean by this?

File: .github/workflows/continuous-integration.yml (hunk 11 above)
Comment (nzpr) [reply .github/workflows/continuous-integration.yml:292]: I guess it is because justificationFollows validation is removed.

File: casper/src/main/scala/coop/rchain/casper/storage/RNodeKeyValueStoreManager.scala (hunk 1 above)
Comment (tgrospic) [reply casper/src/main/scala/coop/rchain/casper/storage/RNodeKeyValueStoreManager.scala:83]: Only couple of full nodes still use this legacy config, but after HF2 we are starting from empty DBs so no need to keep it anymore.

File: casper/src/main/scala/coop/rchain/casper/Casper.scala (hunk 3 above)
Comment (tgrospic) [reply casper/src/main/scala/coop/rchain/casper/Casper.scala:56]: All related data to Approved block will be removed.

Approved block a special block which carries info about signatures of validators for genesis ceremony. This is like a special type of consensus for genesis block which is not based on the stake but on number of signatures.
So all this logic will be replaced by the same finalization algorithm which is used for any other block in the same way for genesis.

File: casper/src/main/scala/coop/rchain/casper/engine/Running.scala (hunk 4 above)
Comment (tgrospic) [reply casper/src/main/scala/coop/rchain/casper/engine/Running.scala:59]: Good catch. I've added a few more commits and removed Casper interface completely. :partying_face: :balloon:

File: casper/src/test/scala/coop/rchain/casper/helper/TestNode.scala (hunk 9 above)
Comment (tgrospic) [reply casper/src/test/scala/coop/rchain/casper/helper/TestNode.scala:381]: It's not used for testing. We can add tests for GraphViz separately.
I didn't want to spend time to fix necessary dependencies to create block API instance here.

File: .github/workflows/continuous-integration.yml (hunk 11 above)
Comment (tgrospic) [reply .github/workflows/continuous-integration.yml:292]: Strangely, the error is `Platform failure: Unable to consume results of system deploy`.
https://github.com/rchain/rchain/runs/6198265818?check_suite_focus=true#step:8:670

File: casper/src/test/scala/coop/rchain/casper/addblock/MultiParentCasperAddBlockSpec.scala (hunk 12 above)
Comment (tgrospic) [reply casper/src/test/scala/coop/rchain/casper/addblock/MultiParentCasperAddBlockSpec.scala]: Without that test will fail with invalid deploy shard ID error.

File: casper/src/main/scala/coop/rchain/casper/BlockStatus.scala (hunk 13 above)
Comment (tgrospic) [reply casper/src/main/scala/coop/rchain/casper/BlockStatus.scala:16]: Yes, I just reuse it for slashable error if some deploy is with invalid shard ID.

File: casper/src/main/scala/coop/rchain/casper/blocks/BlockProcessor.scala (hunk 14 above)
Comment (tgrospic) [reply casper/src/main/scala/coop/rchain/casper/blocks/BlockProcessor.scala]: No, BlockProcessor will manage slashable validation with replay.

Shard will be validated in BlockReceiver so here it must be correct.

File: casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala (hunk 15 above)
Comment (tgrospic) [reply casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala:274]: This was previous code. Do you think it's different? (I'm not sure what is doing :smiley:)
```scala
private def updateLastFinalizedBlock(newBlock: BlockMessage): F[Unit] =
  lastFinalizedBlock.whenA(
    newBlock.body.state.blockNumber % casperShardConf.finalizationRate == 0
  )

def lastFinalizedBlock: F[BlockMessage] =
  for {
    dag          <- blockDag
    blockMessage <- dag.lastFinalizedBlockUnsafe.flatMap(BlockStore[F].getUnsafe)
  } yield blockMessage
```
But it will be removed with the new finalizer.

Comment (nzpr) [reply casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala:274]: After recent PRs with removal of estimator and pure DagRepresentation it just reads finalized block, but particularly here previously `lastFinalizedBlock` method were actually triggering finalization process (which is odd), so just wanted to mention it. 
`But it will be removed with the new finalizer.` 😃

File: casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala [hunk 16]
```diff
@@ -14,56 +14,17 @@ import coop.rchain.casper.merging.BlockIndex
...
-import coop.rchain.metrics.implicits._
 import coop.rchain.metrics.{Metrics, Span}
 import coop.rchain.models.BlockHash._
 import coop.rchain.models.syntax._
 import coop.rchain.models.{BlockHash => _, _}
 import coop.rchain.shared._
 import coop.rchain.shared.syntax._
 
-// format: off
-class MultiParentCasperImpl[F[_]
```
Comment (nzpr) [inline casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala:31]: Oh, such a relief!

**Review by nzpr — APPROVED** [review]
Great!

File: .github/workflows/continuous-integration.yml (hunk 11 above)
Comment (zsluedem) [reply .github/workflows/continuous-integration.yml:292]: We should rewrite these after casper is done. I hope we could implement these slashing tests in scala not Python.

File: casper/src/slowcooker/scala/coop.rchain.casper/HashSetCasperSpecification.scala [hunk 17]
```diff
@@ -1,29 +1,22 @@
...
-import coop.rchain.crypto.PublicKey
-import coop.rchain.models.BlockHash
-import coop.rchain.rholang.interpreter.util.RevAddress
+import coop.rchain.crypto.signatures.Signed
 import monix.eval.Task
 import monix.execution.Scheduler.Implicits.global
 import org.scalacheck._
 import org.scalacheck.commands.Commands
 
+import scala.util.Try
```
Comment (stanislavlyalin) [inline casper/src/slowcooker/scala/coop.rchain.casper/HashSetCasperSpecification.scala:18]: This import is not necessary

File: casper/src/test/scala/coop/rchain/casper/engine/GenesisCeremonyMasterSpec.scala [hunk 18]
```diff
@@ -62,9 +71,7 @@ class GenesisCeremonyMasterSpec extends WordSpec {
           validatorPk
         )
         _               <- EngineCell[Task].read >>= (_.handle(local, blockApproval))
-        blockO          <- blockStore.get1(genesis.blockHash)
-        _               = assert(blockO.isDefined)
-        _               = assert(blockO.contains(genesis))
+        blockO          <- waitForGenesisBlock
```
Comment (stanislavlyalin) [inline casper/src/test/scala/coop/rchain/casper/engine/GenesisCeremonyMasterSpec.scala:74]: Why did the behavior of the code change in this case? :thinking:

**Review by stanislavlyalin — APPROVED** [review]
Great refactoring!
Just a question. Why we changed object BlockAPI to trait BlockApi but Casper trait to MultiParentCasper object here? To be able to remove object functions in future?

**Comment by stanislavlyalin** [comment]
bors merge
